### PR TITLE
Test submission/elson ornelas test

### PR DESCRIPTION
**Pull request by WebTalentTop**

### CODE REVIEW

File: girderformindlogger/api/v1/user.py [hunk 1]
```diff
@@ -845,6 +846,30 @@ def updateUser(
...
+    )
+    def changeUsername(
+        self,
+        login=""
+    ):
+        user = self.getCurrentUser()
+        if len(login):
+            user['login'] = login
+
+        return self._model.save(user)
```
Comment (shnizzedy) [inline girderformindlogger/api/v1/user.py]: > We definitely don't want to expose the whole unfiltered User model, which includes the [userId](https://github.com/ChildMindInstitute/mindlogger-app-backend/issues/153), the [cryptographic salt](https://en.wikipedia.org/wiki/Salt_(cryptography)), and the user's cached applets.

See https://github.com/ChildMindInstitute/mindlogger-app-backend/pull/616#discussion_r367973840, https://github.com/ChildMindInstitute/mindlogger-app-backend/pull/618

File: girderformindlogger/api/v1/user.py [hunk 2]
```diff
@@ -845,6 +846,30 @@ def updateUser(
...
+    @access.user
+    @autoDescribeRoute(
+        Description("Change username.")
+        .param(
+            'login',
+            'Display username',
+            default="",
+            required=False
+        )
+        .errorResponse()
```
Comment (shnizzedy) [inline girderformindlogger/api/v1/user.py]: Two `.errorResponse()`s?

**Comment by shnizzedy** [comment]
Also, I think merging https://github.com/ChildMindInstitute/mindlogger-app-backend/pull/637 into this branch will get the CI tests to pass

File: girderformindlogger/api/v1/applet.py [hunk 3]
```diff
@@ -170,24 +173,31 @@ def assignGroup(self, folder, group, role, subject):
...
+            )
+            thread.start()
+            return({
+                "message": "The applet is being created. Please check back in "
+                        "several mintutes to see it. If you have an email "
+                        "address associated with your account, you will receive "
+                        "an email when your applet is ready."
+            })
+        else:
+            raise ValidationException('An applet with that name already exists.', 'name')
```
Comment (shnizzedy) [inline girderformindlogger/api/v1/applet.py]: This code block appears to be the meat of attempting to resolve #609. 

However, most cases will not be caught by this check.

If the optional param `name` is not provided to https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/api/v1/applet.py#L174

`name` will be inferred from the protocol 

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/models/applet.py#L163-L166

and if an applet with that system name already exists, the new one will be automatically renamed with a number in parentheses following the name.

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/models/applet.py#L91

Finally, there are several semantically keyed names that are preferred to the system name, and this function does not check against any of them:

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/71bff5ddcab3597b92968e18a2e62e07ec609e02/girderformindlogger/constants.py#L104

For example, the "MindLogger Demo" shown in this screenshot has a system `name` of "reprolib:protocols/mindlogger-demo/mindlogger-demo_schema (2)" but `skos:prefLabel` of "MindLogger Demo"

admin screenshot

If I try to create a new "MindLogger Demo", this code lets me. 

```REST
POST /applet?protocolUrl=https%3A%2F%2Fraw.githubusercontent.com%2FReproNim%2Freproschema%2Fmaster%2Fprotocols%2Fmindlogger-demo%2Fmindlogger-demo_schema&name=MindLogger%20Demo
```
`200`
```json
{
  "message": "The applet is being created. Please check back in several mintutes to see it. If you have an email address associated with your account, you will receive an email when your applet is ready."
}
```

If I try to create a new "reprolib:protocols/mindlogger-demo/mindlogger-demo_schema (2)", the code correctly rejects.

```REST
POST /applet?protocolUrl=https%3A%2F%2Fraw.githubusercontent.com%2FReproNim%2Freproschema%2Fmaster%2Fprotocols%2Fmindlogger-demo%2Fmindlogger-demo_schema&name=reprolib%3Aprotocols%2Fmindlogger-demo%2Fmindlogger-demo_schema%20(2)
```
`400`
```json
{
  "field": "name",
  "message": "An applet with that name already exists.",
  "type": "validation"
}
```

Comment (WebTalentTop) [reply girderformindlogger/api/v1/applet.py]: Hi, @shnizzedy. I have a few questions.
- If I input protocol URL already used, creation of the applet should be blocked?
- If name field is not specified, Protocol().preferredName(protocol) will be the applet name.
   If it duplicates, what actions should be taken?

Comment (shnizzedy) [reply girderformindlogger/api/v1/applet.py]: The issue says

> To resolve this issue, 
> 1. [ ] design a strategy for
>     - [ ] checking for any existing applets with the same name as an applet being created or renamed
>     - [ ] if an applet with that name already exists, renaming the new applet without giving the creator of the new applet details of the existing same-name applet
> 2. [ ] implement that strategy

So @webtalenttop, to answer your questions:

> If I input protocol URL already used, creation of the applet should be blocked?

Multiple applets should be able to use the same protocol, so at the very least a user reusing a protocol should have the opportunity to choose a new name. I generally lean toward automation, (eg, "Existing Name (2)" or using something fun like [anonymous Google Docs animals](https://medium.com/@roeder/these-are-all-of-the-anonymous-google-docs-animals-i-have-seen-so-far-8e33e0405993) ("Existing Name: Alligator"))

> If name field is not specified, Protocol().preferredName(protocol) will be the applet name. If it duplicates, what actions should be taken?

The same strategy as whatever you decide for above, ie,  at the very least a user reusing a name should have the opportunity to choose a new name, or an automated renaming strategy.

See [Schema Standardization](https://github.com/ReproNim/reproschema#schema-standardization) for more about the motivations and implementation of protocols in this project.

**Comment by WebTalentTop** [comment]
Hi, @shnizzedy 
I can not run the app at the point of the latest master.
What is the solution?

**Comment by shnizzedy** [comment]
> Said the email will be sent soon but no result.

@WebTalentTop what's the output of your terminal? It should show it slogging away loading JSON-LD documents, which can take a long time.

Here's a visualization of the primary protocol in development, where each node is a document that takes 1‒2 seconds to load:
https://raw.githubusercontent.com/hotavocado/HBN_EMA_NIMH2/master/protocols/EMA_HBN_NIMH2/EMA_HBN_NIMH2_schema

If you can see your shell outputs, you should be able to see if/when the applet succeeds in building regardless of whether your localhost succeeds in sending your email or if your email client [greylists](https://en.wikipedia.org/wiki/Greylisting) the message.

**Comment by WebTalentTop** [comment]
```
Running in mode: production
Traceback (most recent call last):
  File "/Users/pro/girder_env/bin/girderformindlogger", line 11, in <module>
    load_entry_point('girderformindlogger', 'console_scripts', 'girderformindlogger')()
  File "/Users/pro/girder_env/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/pro/girder_env/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/pro/girder_env/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/pro/girder_env/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/pro/girder_env/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Volumes/REPOS/ChildMind/MindLoggerBackend/mindlogger-app-backend/girderformindlogger/cli/serve.py", line 41, in main
    server.setup(mode)
  File "/Volumes/REPOS/ChildMind/MindLoggerBackend/mindlogger-app-backend/girderformindlogger/utility/server.py", line 154, in setup
    girderWebroot, appconf = configureServer(mode, plugins, curConfig)
  File "/Volumes/REPOS/ChildMind/MindLoggerBackend/mindlogger-app-backend/girderformindlogger/utility/server.py", line 78, in configureServer
    from girderformindlogger.api import api_main
  File "/Volumes/REPOS/ChildMind/MindLoggerBackend/mindlogger-app-backend/girderformindlogger/api/api_main.py", line 5, in <module>
    from .v1 import activity, api_key, applet, assetstore, collection, context,    \
  File "/Volumes/REPOS/ChildMind/MindLoggerBackend/mindlogger-app-backend/girderformindlogger/api/v1/notification.py", line 6, in <module>
    from pyfcm import FCMNotification
ModuleNotFoundError: No module named 'pyfcm'
```

**Comment by shnizzedy** [comment]
Your local environment is missing `pyfcm` is a newer requirement than when I was working on the project. It was added to `requirements.txt` in 866590dc634f64fbbe50e2e09fa0b4b4fb1c44ba: 

https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/866590dc634f64fbbe50e2e09fa0b4b4fb1c44ba/requirements.txt#L46

I'd try (from your Girder for MindLogger base directory):

```sh
pip install -r requirements.txt
pip install -e .
girderformindlogger build
girderformindlogger serve
```

and try the admin panel again.

**Comment by WebTalentTop** [comment]
```
Installing collected packages: urllib3, jmespath, docutils, python-dateutil, botocore, s3transfer, boto3, CherryPy, click, click-plugins, cryptography, dictdiffer, dnspython, decorator, dogpile.cache, filelock, isodate, pyrsistent, zipp, importlib-metadata, jsonschema, MarkupSafe, Mako, numpy, pandas, bcrypt, passlib, pymongo, PyYAML, psutil, pyld, pyOpenSSL, pillow, qrcode, chardet, certifi, idna, requests, simplejson, tzlocal, girderformindlogger
  Found existing installation: CherryPy 18.5.0
    Uninstalling CherryPy-18.5.0:
ERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '_json.py'
Consider using the `--user` option or check the permissions.
```

**Comment by shnizzedy** [comment]
https://github.com/ChildMindInstitute/mindlogger-app-backend/blob/96e6ee8024a681d37e6cb753eb02c7ba00534b84/README.rst#L16-L33

**Comment by WebTalentTop** [comment]
Python process gets stopped unexpectedly with new installation.
I should post my solution and you review it.

models/applet.py

 appletName = name if name is not None and len(name) else Protocol(
        ).preferredName(
            protocol
        )
        existingApplets = list(AppletModel().find({
             'name': name
         })
         if len(existingApplets) > 0 :
             appletName = name + '(' + str(len(existingApplets)) + ')';

        applet = self.createApplet(
            name=name,
            name=appletName,

**Comment by shnizzedy** [comment]
A few notes:

1. ```python
   existingApplets = list(AppletModel().find({
   ```
   `AppletModel` is the class this method is defined in, so `self` would be the way to go here.
2. ```python
   list(AppletModel().find({
   'name': name
   })
   ```
   needs one more `)`.
3. ```python
   applet = self.createApplet(
        name=name,
        name=appletName,
   ```
   defines the argument `name` twice.

Resolving those issues, plugging the code in, building and running, the issue remains unresolved:

![from 3 to 4](https://user-images.githubusercontent.com/5974438/74967994-0e8f1e80-53e8-11ea-81ac-ec27f9a2fe0a.gif)

Also, when I ran this test, the log strangely said

```bash
Your applet, None, has been successfully created. The applet's ID is applet/5e4ed0d257c89670d7f6feaf
```

**Comment by WebTalentTop** [comment]
pip install is failed.

```
ERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '_json.py'
Consider using the `--user` option or check the permissions.```

**Pull request closed without merging**